### ISSUE TITLE
Problem with insertCommentStyle method

### ISSUE THREAD

**Issue by JonCook** [nom-tam-fits/nom-tam-fits#68]
Hi,

In the process of trying to upgrade from 1.06 to the latest snapshot nom-tam-fits-1.14.0-20151227.142206-107.jar I have encountered what seems to be a problem with the insertCommentStyle method.

Take the following code example (it is a very simple example only to demonstrate the problem with this method):
```java
float[][] data = new float[500][500];
Fits f = new Fits();
BasicHDU<?> hdu = FitsFactory.hduFactory(data);
hdu.getHeader().insertCommentStyle("        ", "---------------FITS Data Generator---------------");
hdu.getHeader().insertCommentStyle("        ", "This product is generated by me.");
hdu.getHeader().addValue("NAXIS3", 1000, "Actual number of energy channels");
BufferedFile bf = new BufferedFile("img.fits", "rw");
hdu.write(bf);
f.write(bf);
bf.close();
f.close();
```

In the older versions of the library the comments displayed correctly like so:
```
SIMPLE  =                    T / Java FITS: Fri Jan 08 15:07:29 CET 2016        
BITPIX  =                  -32                                                  
NAXIS   =                    2 / Dimensionality                                 
NAXIS1  =                  500                                                  
NAXIS2  =                  500                                                  
NAXIS3  =                 1000 / Actual number of energy channels               
PCOUNT  =                    0 / No extra parameters                            
GCOUNT  =                    1 / One group                                      
EXTEND  =                    T / Extension permitted
        ---------------FITS Data Generator---------------                
        This product is generated by me.
END
```

In the latest version, one of the comments is lost:
```
SIMPLE  =                    T / Java FITS: Fri Jan 08 15:07:29 CET 2016        
BITPIX  =                  -32                                                  
NAXIS   =                    2 / Dimensionality                                 
NAXIS1  =                  500                                                  
NAXIS2  =                  500                                                  
NAXIS3  =                 1000 / Actual number of energy channels               
PCOUNT  =                    0 / No extra parameters                            
GCOUNT  =                    1 / One group                                      
EXTEND  =                    T / Extension permitted
        This product is generated by me.
END
```
I suspect this has something to do with the way the cards are inserted using an iterator which perhaps can't now support duplicate keys properly?

We insert comments in this way, whether it is right or wrong apparently it is supported by the FITS standard.

Thanks for any help.

**Comment by JonCook** [comment]
I think before an unkeyedIndex was used to generate unique keys for elements entered without an key.

**Comment by ritchieGitHub** [comment]
Yes, the problem was that the empty-key comment style (that is not recommented) is used. In the documentation this was originally intended to make complete blank lines as separators. But you can use them as comments ;-)

unit test included, and new snapshot build, please give it a try

**Comment by ritchieGitHub** [comment]
small note: a nicer version of your provided test code, by using closable resources
```java
 try (Fits f = new Fits()) {
            BasicHDU<?> hdu = FitsFactory.hduFactory(data);
            hdu.getHeader().insertCommentStyle("        ", "---------------FITS Data Generator---------------");
            hdu.getHeader().insertCommentStyle("        ", "This product is generated by me.");
            hdu.getHeader().addValue("NAXIS3", 1000, "Actual number of energy channels");
            f.addHDU(hdu);
            try (BufferedFile bf = new BufferedFile("target/testCommentStyleInput.fits", "rw")) {
                f.write(bf);
            }
        }
```

**Comment by JonCook** [comment]
Yes great. Thanks will give it a go Monday.

**Comment by JonCook** [comment]
Not sure if it is possible to re-open this or if you prefer a new issue but the HIERARCH suffers from the same problem. Thanks.

**Comment by JonCook** [comment]
Do i need to provide any more information?

**Comment by ritchieGitHub** [comment]
For example why does it have to be formated the way you wrote? Is there a "compromize" between the way we do it and the way you need it?

**Comment by JonCook** [comment]
Sorry, not sure what you mean. I need to add multiple HIERARCH keys. Thanks.

**Comment by ritchieGitHub** [comment]
No, all ok I will check that this weekend.

**Comment by JonCook** [comment]
ok, thats great. Thanks for all your help.

**Comment by ritchieGitHub** [comment]
now I notice, HIERARCH should occur only once in the fits file. The meaning is that the HIERARCH system is used. HIERARCH-keys are just like fits keys only separated by blanks and or dots.
So no duplicate keys in that case. Or did I misunderstand your error?

**Comment by JonCook** [comment]
Sorry if I didn't explain it very well but the problem is we must be able to write multiple HIERARCH for example:

```
HIERARCH  key.TYPE='type'                                                       
HIERARCH  key.CREATOR='creator'                                                 
HIERARCH  key.DATE='creationDate'                                               
HIERARCH  key.DESC='description'                                                
HIERARCH  key.INSTRUME='instrument'                                             
HIERARCH  key.MODELNAM='modelName'                                              
HIERARCH  key.DATE-OBS='startDate'                                              
HIERARCH  key.DATE-END='endDate'                                                
HIERARCH  key.FORMATV='formatVersion'
```

It seems this should be possible - https://heasarc.gsfc.nasa.gov/docs/software/fitsio/c/f_user/node28.html and makes up a key part of our system for reading/writing. 

Maybe with the new version of the library it must be done in a different way, this code demonstrates the problem and is a simplified version of how we currently write the HIERARCH keyword.

```java
    private static final StringBuilder HIERARCH = new StringBuilder();
    
    private static void writeHierarchy() throws FitsException, IOException {
        FitsFactory.setLongStringsEnabled(true);
        float[][] data = new float[500][500];
        try (Fits f = new Fits()) {
            BasicHDU<?> hdu = FitsFactory.hduFactory(data);
            addHierachs(hdu.getHeader(), "type", "creator", "creationData", "hello");
            f.addHDU(hdu);
            try (BufferedFile bf = new BufferedFile("testHierachInput.fits", "rw")) {
                f.write(bf);
            }
        }
    }
    
    private static void addHierachs(Header header, String... keys){
        for (String key : keys) {
            addHierarchyParam(header, key, key.toUpperCase());
        }
    }

    private static void addHierarchyParam(Header header, String strHcssKey, String strFitsKey) {
        HIERARCH.setLength(2);
        HIERARCH.append("key.").append(strFitsKey).append("='");
        HIERARCH.append(strHcssKey).append('\'');

        header.insertCommentStyle("HIERARCH", HIERARCH.toString());
    }
```
It is the  same method insertCommentStyle which doesn't allow it? I didn't see a specific method for adding HIERARCH?

Thanks

**Comment by ritchieGitHub** [comment]
Ahhh, that explain s it. commentStyle headers may not include an = character
Your old version probably did not support hierarchy keywords directly, the current does!

Handle them as if they where normal keywords . Just turn on the Hierarchy support. In the FitsFactory.
But see #16 for some specialities.

**Comment by JonCook** [comment]
Ok thanks. Sorry if I'm being a dim but I tried adding:
'''
FitsFactory.setUseHierarch(true);
'''
and changing to a normal keyword
```
header.addValue("HIERARCH", HIERARCH.toString(),"");
```

Still doesn't seem to work?

**Comment by ritchieGitHub** [comment]
FitsFactory.setUseHierarch(true);
hdu.addValue("HIERARCH.MONOCH.WAVELENG",624.996,"monochromator wavelength")

as an example

**Comment by ritchieGitHub** [comment]
But see #16 for some specialities.

**Comment by JonCook** [comment]
ok but it doesn't seem possible to replicate this format
```
HIERARCH  key.TYPE='type'                                                       
HIERARCH  key.CREATOR='creator'                                                 
HIERARCH  key.DATE='creationDate'                                               
HIERARCH  key.DESC='description'                                                
HIERARCH  key.INSTRUME='instrument'                                             
HIERARCH  key.MODELNAM='modelName'                                              
HIERARCH  key.DATE-OBS='startDate'                                              
HIERARCH  key.DATE-END='endDate'                                                
HIERARCH  key.FORMATV='formatVersion'
...
```
We have many FITS files in the wilderness which we need to read in this way and also I prefer if possible to keep this format. Thanks.

**Comment by ritchieGitHub** [comment]
We had this discussion.... Ok lets see if I can create an option for that format. Only a single blank after hierarch is ok?

**Comment by JonCook** [comment]
That should be ok and the extra option would be great. I'm sorry to keep creating you tasks! Thanks again.

**Comment by ritchieGitHub** [comment]
Ok, option is available use:
```java
FitsFactory.setHierarchFormater(new BlankDotHierarchKeyFormatter(2));
```
Creating tasks is good! So we know the need of the "wild" ;-)

If you want to give something back ;-) you could write a paragraph (or more) documentation about  this new setting, that will go directly in the documentation / javadoc.

**Comment by JonCook** [comment]
Yes sure. Shall I just add it to the source code directly? Thanks again. Will give it ago tomorrow.

**Comment by ritchieGitHub** [comment]
A pull request is always very welcome ;-) 

In nom-tam-fits/src/site/markdown you can find exisiting docu's. Wenn you want, you can add some text that would have helped you migrating from 1.06 to the current version.

**Comment by JonCook** [comment]
It seems to work thanks.
```
HIERARCH  key.TYPE= 'type    '                                                  
HIERARCH  key.CREATOR= 'creator'                                                
HIERARCH  key.CREATIONDATA= 'creationData'                                      
HIERARCH  key.HELLO= 'hello  ' 
```
Just one question, the padding is deliberate I mean before the single quote closes? e.g) 'type    '?

And likewise the space after the = 

Thanks again.

**Comment by ritchieGitHub** [comment]
The padding is a left offer of the normal headers where it is "tryed" to align the strings, I will see if I can get that eliminated for HIERARCH keys as it does not make sens there.

**Comment by JonCook** [comment]
ok thanks.

**Comment by ritchieGitHub** [comment]
When I remember correctly there is access to the duplicate cards. I can write in more detail on Friday. (just mobil access at the moment)

**Comment by ritchieGitHub** [comment]
I think you forgot  
```java
FitsFactory.setUseHierarch(true); 
```
could you verify if that is the problem?

**Comment by JonCook** [comment]
I need to take the latest snapshot to test this then?

**Comment by JonCook** [comment]
Yes that was the problem. Sorry about that. I had switched it on for writing but not reading. 

I can also confirm the padding issue is fixed. Thanks.

**Comment by ritchieGitHub** [comment]
Please give me a note when all your test are green. Then I will release 1.14.0

**Comment by JonCook** [comment]
Yes I am happy things are working. Thanks.

**Comment by ritchieGitHub** [comment]
version 1.14.0 is released to central (and in the download area)
